README
PDE version 0.01
================

PDE is a collection of extension and configuration for perl
programming using Emacs.


INSTALLATION

To install PDE, add the following to your ~/.emacs:

     (add-to-list 'load-path "/path/to/pde/")
     (load "pde-load")

DEPENDENCIES

None.

COPYRIGHT AND LICENCE

Copyright (C) 2007, Ye Wenbin

This library is free software; you can redistribute it and/or modify
it under the same terms as Perl itself.


ChangeLog
2008-04-26  Ye Wenbin  <wenbinye@gmail.com>

	* lisp/pde-util.el (pde-module-bounds): new function to detect
	perl module at point.

2008-04-23  Ye Wenbin  <wenbinye@gmail.com>

	* lisp/pde-project.el (pde-project-find-file): fix bug: use bind
	default-directory local instead of cd.

2008-04-20  Ye Wenbin  <wenbinye@gmail.com>

	* lisp/pde-util.el (pde-yaml-dump, pde-yaml-load): Add commands
	for read or dump YAML data.

2008-04-19  Ye Wenbin  <wenbinye@gmail.com>

	* lisp/pde-project.el (pde-project-find-file): New command to find
	file in current project.

2008-04-13  Ye Wenbin  <wenbinye@gmail.com>

	* lisp/compile-dwim.el (compile-dwim-alist): Make win32 happy.

	* lisp/tree-mode.el (tree-mode-button-current-line): fixed it

2008-04-01  Ye Wenbin  <wenbinye@gmail.com>

	* lisp/compile-dwim.el (compile-dwim-calculate-command): Test
	whether current file name is changed when generate new commands.
	(compile-dwim-calculate-command): Make compile-dwim recache when
	change file name

2008-03-19  Ye Wenbin  <wenbinye@gmail.com>

	* lisp/re-builder-x.el (reb-perl-start-process): coding system
	support

	* lisp/tempo-x.el (tempo-x-insert-form): Barf when variable is not
	source before

2008-03-18  Ye Wenbin  <wenbinye@gmail.com>

	* lisp/re-builder-x.el: New library for test regexp

	* lisp/perldoc.el (perldoc-read-module, perldoc-module-ap): find
	not module item, too

2008-03-11  Ye Wenbin  <wenbinye@gmail.com>

	* lisp/inf-perl.el (inf-perl-set-cwd): echo working directory

2008-03-09  Ye Wenbin  <wenbinye@gmail.com>

	* lisp/pde.el (pde-perl-mode-hook): Fix error message

	* Release PDE-v0.2.6

	* lisp/pde-abbv.el (pde-abbv-no-blank): Make some abbrev expand
	without insert whitespace.

2008-02-29  Ye Wenbin  <wenbinye@gmail.com>

	* lisp/pde.el (pde-pod-to-manpage): Fix for search pod from point-min

2008-01-13  Ye Wenbin  <wenbinye@gmail.com>

	* lisp/perldoc.el (perldoc-symbol-type): New function to distinct
	symbols in obarray. Change relate functions to use this.

	* lisp/pde-util.el (pde-apropos-module): New command

2008-01-06  Ye Wenbin  <wenbinye@gmail.com>

	* lisp/pde.el (pde-compilation-buffer-name): fix bug: not 1 but 0

	* lisp/tempo-x.el (tempo-x-snippet-clear): fix bug: clear forms
	when source overlay is deleted

2008-01-01  Ye Wenbin  <wenbinye@gmail.com>

	* lisp/compile-dwim.el and others: change defvar to defcustom
	definition

	* lisp/doc/pde.texi (perlcritic): Add docuemnt for perlcritic

2007-12-31  Ye Wenbin  <wenbinye@gmail.com>

	* lisp/pde.el (pde-ffap-locate): make ffap-highlight correct region

	* lisp/perldoc.el (perldoc-module-ap): recache if the thing at
	point like a module

	* lisp/perlcritic.el: add perlcritic.el

	* lisp/perltidy.el (perltidy-region): check perltidy

	* lisp/pde.el (pde-compilation-buffer-name): Use mode name to
	create buffer.

2007-12-30  Ye Wenbin  <wenbinye@gmail.com>

	* lisp/pde-util.el (pde-search-cpan): New command

	* lisp/perldoc.el (perldoc): add a parameter for module convert

	* lisp/pde-util.el (pde-list-core-modules): fix for view "open"

	* lisp/perldoc.el (perldoc-module-ap): Handle name like package::sub.
	(perldoc-locate-module): locate pod also.
	(perldoc): support for encoding

	* lisp/pde-load.el ("ffap"): let pde-ffap-locate handle everywhere

	* lisp/tree-mode.el (tree-mode-toggle-expand-node): A notify
	function to toggle expanding tree

	* lisp/pde-abbv.el: Add pde-abbv.el

	* lisp/tempo-x.el: Add tempo-x.el

2007-12-29  Ye Wenbin  <wenbinye@gmail.com>

	* lisp/perldoc.el (perldoc-tree-widget): click on toplevel button
	toggle expand node.

	* lisp/contrib/tabbar.el: Move tabbar to contrib directory

	* lisp/perldoc.el (perldoc-tree-theme): Add option for
	perldoc-tree theme

	* Build.PL (ACTION_fakeinstall): add fakeinstall action for elisp file.
	(ACTION_disthtml): change action html to disthtml, so make will success.

2007-12-28  Ye Wenbin  <wenbinye@gmail.com>

	* Build.PL: use create_makefile_pl to create Makefile.PL. Thanks
	Michael G Schwern for give the suggestion.

2007-12-27  Ye Wenbin  <wenbinye@gmail.com>

	* lisp/tree-mode.el (tree-mode-toggle-expand): new command
	available for expand and contract node

	* lisp/pde-load.el (pde-extra-setting): don't use nconc, too
	dangerous

	* lisp/pde.el (pde-perl-mode-hook): Use
	executable-make-buffer-file-executable-if-script-p instead of
	executable-chmod

	* lisp/perldoc.el (perldoc-init): New function to init obarray

	* lib/PDE.pm: Mention other emacs libraries and Perl modules

	* lisp/pde.el (pde-pod-to-manpage): using pde-file-package to
	detect module name.
	(pde-tabbar-register): make function autoload so you can use
	perldoc-tree without enter cperl-mode

	* lisp/template-simple.el (define-template-expander): Easy to
	define new expand function.

	* lisp/perldoc.el (perldoc-pragam-expand): fix typo

	* lisp/template-simple.el (template-kill-ring-save): A command
	help write template in string
	(template-compile): Not handle list of element now
	(template-expand-function): Default function is
	template-tempo-expand now.

	* lisp/templates/TEMPLATE.el.tpl: Change file name in case
	somebody put this to library and become a library TEMPLATE.el.

	* lisp/template-simple.el (template-skeleton-expand)
	(template-tempo-expand): Fix bugs introduced

	* lisp/pde-util.el (pde-generate-loaddefs): For automatic create
	pde-loaddefs.el

	* lisp/pde-load.el (pde-extra-setting): Make transient-mark-mode
	default turn on

2007-12-26  Ye Wenbin  <wenbinye@gmail.com>

	* First Release!


lisp/pde-abbv.el
;;; pde-abbv.el --- Abbrevs for perl

;; Copyright (C) 2007 Free Software Foundation, Inc.
;;
;; Author: Ye Wenbin <wenbinye@gmail.com>
;; Maintainer: Ye Wenbin <wenbinye@gmail.com>
;; Created: 30 Dec 2007
;; Version: 0.01
;; Keywords: abbrev, convenience

;; This program is free software; you can redistribute it and/or modify
;; it under the terms of the GNU General Public License as published by
;; the Free Software Foundation; either version 2, or (at your option)
;; any later version.
;;
;; This program is distributed in the hope that it will be useful,
;; but WITHOUT ANY WARRANTY; without even the implied warranty of
;; MERCHANTABILITY or FITNESS FOR A PARTICULAR PURPOSE.  See the
;; GNU General Public License for more details.
;;
;; You should have received a copy of the GNU General Public License
;; along with this program; if not, write to the Free Software
;; Foundation, Inc., 675 Mass Ave, Cambridge, MA 02139, USA.

;;; Commentary:

;; These abbrevs are available:
;;  newx  -  Create subroutine new for Object module.
;;  dumpx -  expand for Data::Dumper::Dumper.
;;  usetk -  Use Tk
;;  exportx - Write Exporter tags
;;  opend -  expand for build-in function `opendir'. snippet available
;;  openf -  expand for build-in function `open'. snippet available
;;  fromto - expand for Encode::from_to, only snippet available
;;  usegtk - Use Gtk2
;;  useenc - Use Encode
;;  usedump - Use Data::Dumper

;; Put this file into your load-path and the following into your ~/.emacs:
;;   (require 'pde-abbv)

;;; Code:

(eval-when-compile
  (require 'cl))
(require 'tempo-x)

(defcustom pde-abbv-use-snippet t
  "*Non-nil means use snippet instead of standard tempo."
  :type 'boolean
  :group 'pde)

(defvar tempo-perl-tags nil
  "Tempo tags for cperl-mode")

(tempo-define-template
 "perl-new"
 '("sub new {"
   n> "my $_class = shift;"
   n> "my $class = ref $_class || $_class;"
   n> "my $self = {};"
   n> "bless $self, $class;"
   n> p
   n> "return $self;"
   n "}" >)
 "newx"
 "Create subroutine new for Object module."
 'tempo-perl-tags)

(tempo-define-template
 "perl-dump"
 '("print Dumper(" p "), \"\\n\";")
 "dumpx"
 "expand for Data::Dumper::Dumper."
 'tempo-perl-tags)

(tempo-define-template
 "perl-usetk"
 '("use Tk;"
   n> "my $mw = MainWindow->new;"
   n> p 
   n "MainLoop;" >)
 "usetk"
 "Use Tk"
 'tempo-perl-tags)

(tempo-define-template
 "perl-export"
 '("require Exporter;"
   n> "our @ISA = qw(Exporter);"
   n> "our %EXPORT_TAGS = ( 'all' => [ qw( " p ") ] );"
   n> "our @EXPORT_OK = ( @{ $EXPORT_TAGS{'all'} } );"
   n "our @EXPORT = qw(  );" >)
 "exportx"
 "Write Exporter tags"
 'tempo-perl-tags)

;; abbrevs not using snippet
(when (null pde-abbv-use-snippet)
  (tempo-define-template
   "perl-opendir"
   '("opendir(" (pi "Dir handle: " "DIR") ", "
     (p "Directory: " dir t)
     (let ((dir (tempo-lookup-named 'dir)))
       (if (string-match "^[$]" dir)
           dir
         (concat "\"" dir "\"")))
     ") or die \"Can't open directory " (s dir) ": $!\";"
     n>)
   "opend"
   "expand for build-in function `opendir'"
   'tempo-perl-tags)

  (tempo-define-template
   "perl-open"
   '("open(" (pi "File handle: " "FH") ", "
     (p "File: " file t)
     (let ((file (tempo-lookup-named 'file)) mode)
       (when (string-match "^\\([><+]+\\)\\s-*" file)
         (setq mode (match-string 1 file)
               file (substring file (match-end 0)))
         (tempo-save-named 'file file))
       (unless (string-match "^[$]" file)
         (setq file (concat "\"" file "\"")))
       (tempo-save-named 'mode "open")
       (if mode
           (if (string-match ">" mode)
               (tempo-save-named 'mode "create"))
         (setq mode "<"))
       (format "\"%s\", %s" mode file))
     ") or die \"Can't " (s mode) " file " (s file) ": $!\";"
     n>)
   "openf"
   "expand for build-in function `open'"
   'tempo-perl-tags))

;; the same abbrevs using snippet
(when pde-abbv-use-snippet
  (tempo-define-template
   "perl-opendir"
   '((snippet
      "opendir(" (S DIR) ", " (S dir "$dir")
      ") or die \"Can't open directory "
      (F (dir)
         (replace-regexp-in-string "['\"]?\\(.*\\)['\"]" "\\1" dir))
      ": $!\";"
      n>))
   "opend"
   "expand for build-in function `opendir'"
   'tempo-perl-tags)

  (tempo-define-template
   "perl-open"
   '((snippet
      "open("
      (S FH "my $fh" t) ", \"" (S op "<") "\", "
      (S file "$file")
      ") or die \"Can't "
      (F (op)
         (if (string-match ">" op)
             "create" "open"))
      " file "
      (F (file)
         (replace-regexp-in-string "['\"]?\\(.*\\)['\"]" "\\1" file))
      ": $!\";"
      n>))
   "openf"
   "expand for build-in function `open'"
   'tempo-perl-tags)
  (tempo-define-template
   "perl-from2"
   '((snippet
      "from_to(" (S oct "$octets") ", \"" (S from_enc) "\", \"" (S to_enc) "\");"
      n>))
   "fromto"
   "expand for Encode::from_to."
   'tempo-perl-tags)
   )

(defun pde-abbv-no-blank ()
  "Expand abbrev without insert blank."
  (if (string= (this-command-keys) " ") t))
(put 'pde-abbv-no-blank 'no-self-insert t)

;; standard abbrev that don't need tempo
(define-abbrev-table 'cperl-mode-abbrev-table
  '(("usegtk" "use Gtk2 '-init';
use Glib qw(TRUE FALSE);

my $window = Gtk2::Window->new('toplevel');
$window->signal_connect('delete_event' => sub { Gtk2->main_quit; });"
     pde-abbv-no-blank 0)
    ("useenc" "use Encode qw(encode decode from_to);"
     pde-abbv-no-blank 0)
    ("usedump" "use Data::Dumper qw(Dumper);"
     pde-abbv-no-blank 0)
    ("useopt" "use Getopt::Long;
GetOptions();" pde-abbv-no-blank 0)))
  
(tempo-define-template
 "filevar"
 '((pi "Prefix: " comment-start prefix) "Local Variables: ***\n"
   (R (s prefix)
      (pi ("Variable: " ("mode" "coding")) nil var)
      (& var (delete-region recursion-start (point)))
      ": " (p "Value: ")
      " ***\n")
   (s prefix) "End: ***\n")
 "filevarx")

(provide 'pde-abbv)
;;; pde-abbv.el ends here
